combat-trainer.lic

```diff
@@ -1998,6 +2017,181 @@ class TrainerProcess
     DRRoom.dead_npcs.size > dead_count
   end
 
+  def visible_bodies
+    result = []
+    all_bodies = @constellations
+    case DRC.bput('observe sky', 'The following heavenly bodies are visible:', "That's a bit hard to do while inside")
+    when "That's a bit hard to do while inside"
+      echo('Must be outdoors to observe sky')
+      @training_abilities.delete('astro')
+      return
+    end
+    until (line = get?) =~ /^Roundtime/i
+      result << all_bodies.find { |body| /\b#{body['name']}\b/i =~ line && line !~ /below the horizon/ }
+    end
+    result.compact.select { |data| data['circle'] <= DRStats.circle }
+  end
+
+  def check_pools
+    pools = {
+      'lore' => 0,
+      'magic' => 0,
+      'survival' => 0,
+      'offensive combat' => 0,
+      'defensive combat' => 0,
+      'future events' => 0
+    }
+    pools_to_size = {
+      /You have no understanding of the celestial influences over/ => 0,
+      /You have a feeble understanding of the celestial influences over/ => 1,
+      /You have a weak understanding of the celestial influences over/ => 2,
+      /You have a fledgling understanding of the celestial influences over/ => 3,
+      /You have a modest understanding of the celestial influences over/ => 4,
+      /You have a decent understanding of the celestial influences over/ => 5,
+      /You have a significant understanding of the celestial influences over/ => 6,
+      /You have a potent understanding of the celestial influences over/ => 7,
+      /You have an insightful understanding of the celestial influences over/ => 8,
+      /You have a powerful understanding of the celestial influences over/ => 9,
+      /You have a complete understanding of the celestial influences over/ => 10
+    }
+    waitrt?
+    DRCT.retreat
+    bput('predict state all', 'roundtime')
```

Comment:
❌ Based on the logs we've seen, I think you need to put this in a loop of some sort and handle re-retreating.

```
[combat-trainer]>predict state all
>
The Xala'shar magus closes to pole weapon range on you!
>
A hulking Xala'shar thrall strolls into view with a nervous gait, scanning the area with wide eyes.
>
A hulking Xala'shar magus spreads its arms and begins chanting in an ancient tongue, it robes thrashing about wildly as if surrounded in an ethereal wind.
 
The Xala'shar magus begins to advance on you!
The Xala'shar magus is still a distance away from you and is closing steadily.
>
You are far too occupied by present matters to immerse yourself in matters of the future.
```